org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java

```diff
@@ -40,6 +42,9 @@ public final class ParentProcessWatcher implements Runnable, Function<MessageCon
 
 	public ParentProcessWatcher(LanguageServer server ) {
 		this.server = server;
+		if (ProcessHandle.current().parent().isPresent()) {
+			this.server.setParentProcessId(ProcessHandle.current().parent().get().pid());
+		}
```

Comment:
Is there a case where having it set in `initialize()` isn't soon enough ? Or is this just a good approach in case it's ever needed earlier.

Reply:
We try to save the workspace if it is possible.

Reply:
But all this does is set the parent process before `initialize()` is processed on the language server. Is there any reason we need to do this ? From what I could tell, when this bug happens, the parent process has already been set on the language server so there isn't a need for this.

Reply:
> But all this does is set the parent process before initialize() is processed on the language server

See https://github.com/eclipse/eclipse.jdt.ls/pull/2303/commits/c9d1eae9909a901841187f2d96e33c12ab9de370#diff-3ddfee476d8323771f9164fb4f58007ae274217f5c51c3341da8943e080e278fR57 and https://github.com/eclipse/eclipse.jdt.ls/pull/2303/commits/c9d1eae9909a901841187f2d96e33c12ab9de370#diff-3f0025a2c8288df8dde3123363e22073f3e6909a7ba99eb31faba6fb5daacb45R164


> Is there any reason we need to do this ? From what I could tell, when this bug happens, the parent process has already been set on the language server so there isn't a need for this.

We know the parent process earlier. If Java LS hangs before calling LanguageServer.setParentProcesId(), the parent process watcher will still exit Java LS.